### ISSUE TITLE
LSTM/RNN

### ISSUE THREAD

**Issue by nyanp** [nyanp/tiny-cnn#210]
Since we can handle generic graph-networks now, it's time to implement LSTM/RNN. Milestones:

- [ ] Add a basic LSTM cell (equivalent to TF's BasicLSTMCell)
- [ ] Modify network class to handle sequential-loss
- [ ] Add utilities functions to for NLP tasks

If no one have a plan to implement it, I'll do it. cc: @bhack @edgarriba @Wangyida

**Comment by bhack** [comment]
@nyanp You can go ahead with this. Probably we will not have any accelerated implementation with actual status of integrated back ends (NNPACK/LIBDNN) so it think this will  be slow to train. Take a look also to [TF rnn API](https://www.tensorflow.org/versions/r0.9/api_docs/python/nn.html#recurrent-neural-networks) /cc @naibaf7

**Comment by naibaf7** [comment]
@bhack 
There is a RNN/LSTM kernel set for OpenCL in OpenCL Caffe, but it's not terribly fast (yet).

**Comment by bhack** [comment]
@naibaf7 Do you plan to port this and other kernels standalone in libdnn? As you can see in https://github.com/tensorflow/tensorflow/issues/1786 cudnn it is forcing a standard defacto API but don't seems frameworks are so fast to converge on that or likes lack of flexibility.

**Comment by zhangqianhui** [comment]
very good

**Comment by naibaf7** [comment]
@bhack 
Yes I mean as a basic I can just include some more OpenCL kernels into libDNN that provide a working solution. But to optimize them all for high performance, I think I don't have time to do that alone... and there are some big things coming up, like FP16 support on the OpenCL Caffe branch.

**Comment by bhack** [comment]
I think that it is better to extend libdnn API and  interface that to caffe. If this doesn't increase to much  the effort could help to attract more contributors. @hughperkins could contribute at some point some really fast kernels that it is porting from neon. I think that the Opencl space in deep learning it is too small to be fragmented and libdnn could be the right, indipendent, collaboration repository (also for Intel that actually is contributing  to the caffe Opencl  branch).

**Comment by Thanh-Binh** [comment]
For RNN and specially Gated Recurrent Unit, I found a interesting implementation
https://github.com/yandex/faster-rnnlm
Maybe we can collaborate with him or find somethings useful for tiny_cnn?

**Comment by zhangqianhui** [comment]
I think the tiny-cnn should facus on the embedded platform and find self's point.

**Comment by zhangqianhui** [comment]
if add the lstm or rnn , what will the name of tiny-cnn  change?

**Comment by nyanp** [comment]
@bhack 
Ok I'll go. In first version, I'll only provide a straightforward version running on cpu. Training time will be slow, but it is not so critical for tiny-cnn.

**Comment by nyanp** [comment]
@Thanh-Binh 
Very interesting, but portability&header-only is more important than fast training. I don't want to introduce multiple 3rd-party GPU libraries...

**Comment by nyanp** [comment]
@zhangqianhui 
> I think the tiny-cnn should facus on the embedded platform and find self's point.

What feature do you think we need to focus on, for embedded users?

> if add the lstm or rnn , what will the name of tiny-cnn change?

Yeah probably we should change its name... when we get LSTM/RNN, it will be neither tiny, nor (only) cnn :)

**Comment by bhack** [comment]
Or tinydnn. @nyanp From the point of view of tensor what kind of rank do you plan to support?

**Comment by nyanp** [comment]
Except for GSoC's scope, I want to support these features (order by my priority):

- Batch Normalization
- LSTM/RNN
- TensorFlow importer
- Model Serialization/Deserialization in json format or caffe compatible format (#188)
- Adding other major layers to support more models
    - There are no clear priority between these layers, but I think following layers are worth supporting : preprocessing(GCN/LCN), residual unit, prelu, word-embedding. Please tell me if you have more important layers we should get. 

I also want to support these small features:
- More application examples for embedded users (Face Recognition on raspberry pi, etc)
- Functional style API for graph construction
- Prediction-only mode

**Comment by bhack** [comment]
@nyanp Nice plan but I mean more generally what tensor rank we want to support: ndpooling, ndconvolution or cover only 3d kernels ops (plus the batch dimension)?

**Comment by nyanp** [comment]
@bhack 
Ah I completely misread... I don't know an example to use more than 3-dimentional kernel ops. Could you tell me one of them?

**Comment by bhack** [comment]
@nyanp https://github.com/BVLC/caffe/pull/3983

**Comment by edgarriba** [comment]
@nyanp from my side apart from OpenCL integration, I'll go for the TF loader during the GSOC.

**Comment by Wangyida** [comment]
@edgarriba Nice to have TF loader here!

**Comment by nyanp** [comment]
@bhack 
Thanks for the pointer, covering 3d kernels would be enough for us. I don't know any major paper which utilize 4d plus batch dimension.

**Comment by nyanp** [comment]
@edgarriba 
Awesome! Sure, you can go ahead :+1:

**Comment by bhack** [comment]
As reference of some new archs.. http://distill.pub/2016/augmented-rnns/

**Comment by hughperkins** [comment]
@bhack Thats a pretty nice summary you found there :-)

**Comment by bhack** [comment]
Another little api overview http://mxnet.io/architecture/rnn_interface.html

**Comment by nyanp** [comment]
It's a shame that I don't have enough time to implement LSTM/RNN now. Any contributions are welcome!

**Comment by weisunding** [comment]
@nyanp Tiny-DNN is awsome with small footprint, any plan to implement LSTM/RNN, or any plan to integrate the https://github.com/tmbdev/clstm (small C++ LSTM) implement?

**Comment by Randl** [comment]
@weisunding It might be implemented at GSoC

**Comment by weisunding** [comment]
@Randl Thanks, I will stay tune to the LSTM/RNN implement.
I am working on stocks/futures deep q-learning, converting the time series into pictures is a little bit awkward.

**Comment by Toby237** [comment]
An implementation of LSTM/RNN would be so great. I'll also stay tuned for that :)

**Comment by srk97** [comment]
Does tiny-dnn support residual connections as of now?

**Comment by prlz77** [comment]
Hi, I have posted a message on google [groups](https://groups.google.com/forum/#!topic/opencv-gsoc-2017/_Diliw_hns4) for implementing RNNs for GSOC2017. As @nyanp says, this implies allowing for multiple input/outputs on the graph, as well as modifying the fit/train functions, and adding examples. As I have seen in the code, there has been an attempt to allow for multiple inputs/outputs in the graph but there are some confusing comments like [this](https://github.com/tiny-dnn/tiny-dnn/blob/2ea371108b47785b4959ccf1226a728bf4199569/tiny_dnn/nodes.h#L363) not implemented, and [this](https://github.com/tiny-dnn/tiny-dnn/blob/2ea371108b47785b4959ccf1226a728bf4199569/tiny_dnn/nodes.h#L145) one too.

To my understanding, with an already implemented graph, it is only matter of creating multiple input/loss nodes in that graph, and properly handling them in the network, am I right?

**Comment by bhack** [comment]
@prlz77 You are welcome. It is really hard to design a good RNN api. See https://github.com/fchollet/keras/issues/3692 and other kind of issues in other frameworks. So seems that also for CUDNN it was hard to design a good acelleration API for rnn components.

**Comment by prlz77** [comment]
@bhack Thank you for the interest. I agree with you it is difficult to design a good RNN API, and that is why I think my contribution would be a good starting point for at least supporting sequential graphs and leave optimization for future contributions. In fact, I am used to work with https://github.com/Element-Research/rnn, without the CUDNN backend, and it is fast enough for most purposes.

**Comment by bhack** [comment]
I think that we could have problem soon or later that other solution are already trying to solve: Dynet, pytorch, tensorflow fold. See https://www.reddit.com/r/MachineLearning/comments/5otdwl/d_tensorflow_vs_pytorch_why_is_dynamic_better/

**Comment by prlz77** [comment]
@bhack every time I have had to implement encoder-decoder and hard attention models in TensorFlow, I have had a hard time due to the static graph implementation, in this sense dynamic is better, although static allows for better optimization. However, I think this is a more transcendental decision on tiny-dnn on which I have no voice since I am new here. As I see it, for now we could avoid this "more advanced" models, and center our attention on making tiny-dnn able to make a prediction on, for instance, n fixed video frames. I think just this "small" improvement would attract the attention of big sectors that need for embedded systems, such like autonomous driving. It would also be easy to adapt it to other kind of sequential data, as long as we could fix the input and output sequence size...

**Comment by edgarriba** [comment]
I agree with @prlz77 that switching the whole library to a dynamic graph model might need a huge effort probably a bit incompatible in time with the current proposal. Maybe @nyanp could give their opinion since he already has initial ideas about RNNs. As a first approach I don't think it's a bad idea to start with fixed  sequence sizes.

**Comment by bhack** [comment]
I don't think dynamic could be handle to gsoc.. but just to be aware that specially rnn introduce this topic.

**Comment by edgarriba** [comment]
@prlz77 maybe you could introduce that topic in your proposal and take it into account during the API design, just in case that in a future we decide to switch to a dynamic model.

**Comment by prlz77** [comment]
@edgarriba @bhack thank you, indeed your comments help to make a better proposal.

**Comment by prlz77** [comment]
I have uploaded a draft, if any mentor is interested I will be happy to  receive some feedback :)

**Comment by prlz77** [comment]
Now that we have a basic rnn cell implemented, I am implementing a wrapper to propagate steps trough time. This has two important implications: 

1. the layer must generate some outputs that are not going to be propagated in network depth but in time.
2. the state must be fed as an `edge` so that the optimizer can update the recurrent weight matrices.

The solution to this has been to use `aux` tensors, since they accomplish both properties.

**The problem** is that edges are not kept in the layer, thus making state persistence more difficult (for instance copying the last timestep to the first one to continue with a truncated sequence). Now I am just saving a copy of the last timestep state in the wrapper to handle it but it does not seem very elegant to me. Can you think of a better solution?

Another issue is that the recurrent wrapper must receive a cell to wrap. Before @beru told me that it is bad for cereal, I was just doing: 
```c++
template <class Cell>
class recurrent_layer: public Cell
```
Now I am passing a cell by reference to the constructor. In order to do it, I must use polymorphism to accept different kind of cells. So I have made all the specific cells to inherit from a `cell` abstract class, do you agree with this procedure?
